src/BugsnagUnity/Clients/Client.cs

```diff
@@ -0,0 +1,223 @@
+        });
+      }
+    }
+
+    public void BeforeNotify(Middleware middleware)
+    {
+      lock (MiddlewareLock)
+      {
+        Middleware.Add(middleware);
+      }
+    }
+
+    public void Notify(System.Exception exception)
+    {
+      Notify(exception, null);
+    }
+
+    public void Notify(System.Exception exception, Middleware callback)
+    {
+      Notify(exception, HandledState.ForHandledException(), callback);
+    }
+
+    public void Notify(System.Exception exception, Severity severity)
+    {
+      Notify(exception, severity, null);
+    }
+
+    public void Notify(System.Exception exception, Severity severity, Middleware callback)
+    {
+      Notify(exception, HandledState.ForUserSpecifiedSeverity(severity), callback);
+    }
+
+    void Notify(System.Exception exception, HandledState handledState, Middleware callback)
+    {
+      Notify(new Exceptions(exception).ToArray(), handledState, callback);
+    }
+
+    void Notify(Exception[] exceptions, HandledState handledState, Middleware callback, LogType? logType = null)
+    {
+      var user = new User { Id = User.Id, Email = User.Email, Name = User.Name };
+      var app = new App(Configuration)
+      {
+        InForeground = InForeground,
+        DurationInForeground = Stopwatch.Elapsed,
+      };
+      NativeClient.PopulateApp(app);
+      var device = new Device();
+      NativeClient.PopulateDevice(device);
+      var metadata = new Metadata();
+      NativeClient.PopulateMetadata(metadata);
+
+      foreach (var item in Metadata)
+      {
+        metadata.Add(item.Key, item.Value);
+      }
+
+      var @event = new Payload.Event(
+        Configuration.Context,
+        metadata,
+        app,
+        device,
+        user,
+        exceptions,
+        handledState,
+        Breadcrumbs.Retrieve(),
+        SessionTracking.CurrentSession);
+      var report = new Report(Configuration, @event);
+
+      if (report.Configuration.ReleaseStage != null
+          && report.Configuration.NotifyReleaseStages != null
+          && !report.Configuration.NotifyReleaseStages.Contains(report.Configuration.ReleaseStage))
+      {
+        report.Ignore();
+      }
+
+      lock (MiddlewareLock)
+      {
+        foreach (var middleware in Middleware)
+        {
+          try
+          {
+            middleware(report);
+          }
+          catch (System.Exception)
+          {
+          }
+        }
+      }
+
+      try
+      {
+        callback?.Invoke(report);
+      }
+      catch (System.Exception)
+      {
+      }
+
+      if (!report.Ignored)
+      {
+        Send(report);
+
+        Breadcrumbs.Leave(Breadcrumb.FromReport(report));
+
+        SessionTracking.CurrentSession?.AddException(report);
+      }
+    }
+
+    public void SetApplicationState(bool inFocus)
+    {
+      if (inFocus)
+      {
+        Stopwatch.Start();
+
+        if (Configuration.AutoCaptureSessions)
+        {
+          SessionTracking.StartSession();
+        }
```

Comment:
Is there any possibility that the Unity callbacks i.e. `OnApplicationFocus` could be called multiple times? If so, this could lead to unwanted sessions.

Reply:
It could indeed be called multiple times, each time though I believe we would want a new session